azurerm/internal/services/keyvault/key_vault_managed_storage_account.go

```diff
@@ -0,0 +1,291 @@
+	}
+
+	keyVaultBaseUrl, err := keyVaultsClient.BaseUriForKeyVault(ctx, *keyVaultId)
+	if err != nil {
+		return fmt.Errorf("looking up Base URI for Managed Storage Account %q in %s: %+v", name, *keyVaultId, err)
+	}
+
+	if d.IsNewResource() {
+		existing, err := client.GetStorageAccount(ctx, *keyVaultBaseUrl, name)
+		if err != nil {
+			if !utils.ResponseWasNotFound(existing.Response) {
+				return fmt.Errorf("checking for presence of existing Managed Storage Account %q (Key Vault %q): %s", name, *keyVaultBaseUrl, err)
+			}
+		}
+
+		if existing.ID != nil && *existing.ID != "" {
+			return tf.ImportAsExistsError("azurerm_key_vault_managed_storage_account", *existing.ID)
+		}
+	}
+
+	t := d.Get("tags").(map[string]interface{})
+
+	parameters := keyvault.StorageAccountCreateParameters{
+		ResourceID:         utils.String(d.Get("storage_account_id").(string)),
+		ActiveKeyName:      utils.String(d.Get("storage_account_key").(string)),
+		AutoRegenerateKey:  utils.Bool(d.Get("auto_regenerate_key").(bool)),
+		RegenerationPeriod: utils.String(d.Get("regeneration_period").(string)),
+		Tags:               tags.Expand(t),
+	}
+
+	if _, err := client.SetStorageAccount(ctx, *keyVaultBaseUrl, name, parameters); err != nil {
+		return fmt.Errorf("creating %s: %+v", name, err)
+	}
+
+	if resp, err := client.SetStorageAccount(ctx, *keyVaultBaseUrl, name, parameters); err != nil {
+		// In the case that the Storage Account already exists in a Soft Deleted / Recoverable state we check if `recover_soft_deleted_key_vaults` is set
+		// and attempt recovery where appropriate
+		if meta.(*clients.Client).Features.KeyVault.RecoverSoftDeletedKeyVaults && utils.ResponseWasConflict(resp.Response) {
+			recoveredStorageAccount, err := client.RecoverDeletedStorageAccount(ctx, *keyVaultBaseUrl, name)
+			if err != nil {
+				return err
+			}
+			log.Printf("[DEBUG] Recovering Managed Storage Account %q with ID: %q", name, *recoveredStorageAccount.ID)
+			// We need to wait for consistency, recovered Key Vault Child items are not as readily available as newly created
+			if secret := recoveredStorageAccount.ID; secret != nil {
+				stateConf := &resource.StateChangeConf{
+					Pending:                   []string{"pending"},
+					Target:                    []string{"available"},
+					Refresh:                   keyVaultChildItemRefreshFunc(*secret),
+					Delay:                     30 * time.Second,
+					PollInterval:              10 * time.Second,
+					ContinuousTargetOccurence: 10,
+					Timeout:                   d.Timeout(schema.TimeoutCreate),
+				}
+
+				if _, err := stateConf.WaitForState(); err != nil {
+					return fmt.Errorf("Error waiting for Managed Storage Account %q to become available: %s", name, err)
+				}
+				log.Printf("[DEBUG] Managed Storage Account %q recovered with ID: %q", name, *recoveredStorageAccount.ID)
+
+				_, err := client.SetStorageAccount(ctx, *keyVaultBaseUrl, name, parameters)
+				if err != nil {
+					return err
+				}
```

Comment:
we can combine these lines? and wrap this error
```suggestion
				if _, err := client.SetStorageAccount(ctx, *keyVaultBaseUrl, name, parameters); err != nil {
					return err
				}
```